distributed/deploy/subprocess.py

```diff
@@ -0,0 +1,124 @@
+from distributed import Scheduler
+from distributed.deploy.spec import ProcessInterface, SpecCluster
+from distributed.deploy.utils import nprocesses_nthreads
+from distributed.worker_memory import parse_memory_limit
+
+logger = logging.getLogger(__name__)
+
+
+class SubprocessWorker(ProcessInterface):
+    scheduler: str
+    worker_class: str
+    worker_options: dict
+    name: str | None
+    process: asyncio.subprocess.Process | None
+
+    def __init__(
+        self,
+        scheduler: str,
+        worker_class: str = "distributed.Nanny",
+        name: str | None = None,
+        worker_options: dict | None = None,
+        **kwargs: Any,
+    ) -> None:
+        self.scheduler = scheduler
+        self.worker_class = worker_class
+        self.name = name
+        self.worker_options = copy.copy(worker_options or {})
+        self.process = None
+        logger.info(kwargs)
+        super().__init__(**kwargs)
+
+    async def start(self) -> None:
+        self.process = await asyncio.create_subprocess_exec(
+            "dask",
+            "spec",
+            self.scheduler,
+            "--spec",
+            json.dumps(
+                {0: {"cls": self.worker_class, "opts": {**self.worker_options}}}
+            ),
+        )
+        await super().start()
+
+    async def close(self) -> None:
+        if self.process and self.process.returncode is None:
+            self.process.kill()
+            await self.process.wait()
+        self.process = None
+        await super().close()
+
+
+def SubprocessCluster(
+    host: str | None = None,
+    scheduler_port: int = 0,
+    scheduler_options: dict | None = None,
+    dashboard_address: str = ":8787",
+    worker_class: str = "distributed.Nanny",
+    n_workers: int | None = None,
+    threads_per_worker: int | None = None,
+    worker_dashboard_address: str | None = None,
+    worker_options: dict | None = None,
+    silence_logs: int = logging.WARN,
+    **kwargs: Any,
+) -> SpecCluster:
+    if not host:
+        host = "127.0.0.1"
+    worker_options = worker_options or {}
+    scheduler_options = scheduler_options or {}
+
+    if n_workers is None and threads_per_worker is None:
+        n_workers, threads_per_worker = nprocesses_nthreads()
+    if n_workers is None and threads_per_worker is not None:
+        n_workers = max(1, CPU_COUNT // threads_per_worker)
+    if n_workers and threads_per_worker is None:
+        # Overcommit threads per worker, rather than undercommit
+        threads_per_worker = max(1, int(math.ceil(CPU_COUNT / n_workers)))
+    if n_workers and "memory_limit" not in worker_options:
+        worker_options["memory_limit"] = parse_memory_limit(
+            "auto", 1, n_workers, logger=logger
+        )
+    assert n_workers
+
+    scheduler_options.update(
+        {
+            "host": host,
+            "port": scheduler_port,
+            "dashboard": dashboard_address is not None,
+            "dashboard_address": dashboard_address,
+        }
```

Comment:
These keys might already be in `scheduler/worker_options`.  Maybe we want to use the `dict.setdefault` method instead